Gem/Code/Source/Components/CharacterComponent.cpp

```diff
@@ -10,9 +10,76 @@
 #include <AzFramework/Physics/Character.h>
 #include <Multiplayer/Components/NetworkTransformComponent.h>
 #include <Multiplayer/NetworkTime/INetworkTime.h>
+#include <PhysXCharacters/API/CharacterController.h>
+#include <PhysX/PhysXLocks.h>
+#include <PhysX/Utils.h>
+#include <Source/Components/NetworkRigidBodyComponent.h>
 
 namespace MultiplayerSample
 {
+    bool CollisionLayerBasedControllerFilter(const physx::PxController& controllerA, const physx::PxController& controllerB)
+    {
+        PHYSX_SCENE_READ_LOCK(controllerA.getActor()->getScene());
+        physx::PxRigidDynamic* actorA = controllerA.getActor();
+        physx::PxRigidDynamic* actorB = controllerB.getActor();
+
+        if (actorA && actorA->getNbShapes() > 0 && actorB && actorB->getNbShapes() > 0)
+        {
+            physx::PxShape* shapeA = nullptr;
+            actorA->getShapes(&shapeA, 1, 0);
+            physx::PxFilterData filterDataA = shapeA->getSimulationFilterData();
+            physx::PxShape* shapeB = nullptr;
+            actorB->getShapes(&shapeB, 1, 0);
+            physx::PxFilterData filterDataB = shapeB->getSimulationFilterData();
+            return PhysX::Utils::Collision::ShouldCollide(filterDataA, filterDataB);
+        }
+
+        return true;
+    }
+
+    physx::PxQueryHitType::Enum CollisionLayerBasedObjectPreFilter(
+        const physx::PxFilterData& filterData,
+        const physx::PxShape* shape,
+        const physx::PxRigidActor* actor,
+        [[maybe_unused]] physx::PxHitFlags& queryFlags)
+    {
+        // non-kinematic dynamic bodies should not impede the movement of the character
+        if (actor->getConcreteType() == physx::PxConcreteType::eRIGID_DYNAMIC)
+        {
+            const physx::PxRigidDynamic* rigidDynamic = static_cast<const physx::PxRigidDynamic*>(actor);
+
+            bool isKinematic = (rigidDynamic->getRigidBodyFlags() & physx::PxRigidBodyFlag::eKINEMATIC);
+            if (isKinematic)
+            {
+                const PhysX::ActorData* actorData = PhysX::Utils::GetUserData(rigidDynamic);
+                if (actorData)
+                {
+                    const AZ::EntityId entityId = actorData->GetEntityId();
+
+                    if (NetworkRigidBodyRequestBus::FindFirstHandler(entityId) != nullptr)
```

Comment:
You also could just try constructing a network entity handle.. if that succeeds then it's a net bound entity, if it doesn't then it's not. Also we should make sure the entity specifically has a net rigid body component on it, and not a net character component. We should also probably make net character incompatible with net rigid body so we can't put both on the same entity.

Reply:
The check here is not if it's net bound or not. We can have net bound kinematic entities that will need to block the characters. 
The check here is for the kinematic rigid bodies that are dynamic on the server (i.e. they are not kinematic from the gameplay perspective). 
That's why we're specifically checking for NetworkRigidBodyRequestBus::FindFirstHandler to see if this entity is a dynamic rigid body on the server.
Good point about making components incompatible!